pkg/trojan/message.go

```diff
@@ -0,0 +1,283 @@
+	ErrVarLenTargets = errors.New("target list cannot have targets of different length")
+
+	ErrInvalidHasher = errors.New("hasher pool not initialized")
+)
+
+// NewTopic creates a new Topic variable with the given input string
+// the input string is taken as a byte slice and hashed
+func NewTopic(topic string) Topic {
+	var tpc Topic
+	t, err := crypto.LegacyKeccak256([]byte(topic))
+	if err != nil {
+		return tpc
+	}
+	copy(tpc[:], t)
+	return tpc
+}
+
+func hashFunc() hash.Hash {
+	return sha3.NewLegacyKeccak256()
+}
+
+var hasher *bmtlegacy.Hasher
+
+// NewMessage creates a new Message variable with the given topic and payload
+// it finds a length and nonce for the message according to the given input and maximum payload size
+func NewMessage(topic Topic, payload []byte) (Message, error) {
+	if len(payload) > MaxPayloadSize {
+		return Message{}, ErrPayloadTooBig
+	}
+
+	// get length as array of 2 bytes
+	payloadSize := uint16(len(payload))
+	lengthBuf := make([]byte, 2)
+	binary.BigEndian.PutUint16(lengthBuf, payloadSize)
+
+	// set random bytes as padding
+	paddingLen := MaxPayloadSize - payloadSize
+	padding := make([]byte, paddingLen)
+	if _, err := rand.Read(padding); err != nil {
+		return Message{}, err
+	}
+
+	// create new Message var and set fields
+	m := new(Message)
+	copy(m.length[:], lengthBuf)
+	m.Topic = topic
+	m.Payload = payload
+	m.padding = padding
+
+	hashPool := bmtlegacy.NewTreePool(hashFunc, swarm.Branches, bmtlegacy.PoolSize)
+	hasher = bmtlegacy.New(hashPool)
+
+	return *m, nil
+}
+
+// Wrap creates a new trojan chunk for the given targets and Message
+// a trojan chunk is a content-addressed chunk made up of span, a nonce, and a payload which contains the Message
+// the chunk address will have one of the targets as its prefix and thus will be forwarded to the neighbourhood of the recipient overlay address the target is derived from
+func (m *Message) Wrap(targets Targets) (swarm.Chunk, error) {
+	if err := checkTargets(targets); err != nil {
+		return nil, err
+	}
+
+	span := make([]byte, 8)
+	// 4064 bytes for trojan message + 32 bytes for nonce = 4096 bytes as payload for resulting chunk
+	binary.LittleEndian.PutUint64(span, swarm.ChunkSize)
+
+	return m.toChunk(targets, span)
+}
+
+// Unwrap creates a new trojan message from the given chunk payload
+// this function assumes the chunk has been validated as a content-addressed chunk
+// it will return the resulting message if the unwrapping is successful, and an error otherwise
+func Unwrap(c swarm.Chunk) (*Message, error) {
+	d := c.Data()
+
+	// unmarshal chunk payload into message
+	m := new(Message)
+	// first 40 bytes are span + nonce
+	err := m.UnmarshalBinary(d[40:])
+
+	return m, err
+}
+
+// IsPotential returns true if the given chunk is a potential trojan
+func IsPotential(c swarm.Chunk) bool {
+	// chunk must be content-addressed to be trojan
```

Comment:
We should defer to the validator for this check (in that case type is no longer needed).

Reply:
Yeah, we will do trojan validator in a later PR, until then for testcases this can be used. Until then, this need not be exported and should use export_test instead.